core/src/main/java/org/hisp/dhis/android/core/arch/api/internal/KtorPlugins.kt

```diff
@@ -25,33 +25,29 @@
  *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-package org.hisp.dhis.android.core.arch.api.internal
-
-import okhttp3.Interceptor
-import okhttp3.Response
-import org.hisp.dhis.android.core.arch.api.authentication.internal.PasswordAndCookieAuthenticator.Companion.LOCATION_KEY
-import java.io.IOException
 
-internal class ServerURLVersionRedirectionInterceptor : Interceptor {
+package org.hisp.dhis.android.core.arch.api.internal
 
-    @Throws(IOException::class)
-    override fun intercept(chain: Interceptor.Chain): Response {
-        val request = chain.request()
-        var response = chain.proceed(request)
+import UserAgentPlugin
+import io.ktor.client.HttpClientConfig
+import io.ktor.client.plugins.HttpTimeout
+import org.hisp.dhis.android.core.D2Configuration
+import org.hisp.dhis.android.core.arch.api.authentication.internal.ParentAuthenticatorPlugin
 
-        var redirects = 0
-        while (response.isRedirect && redirects <= MaxRedirects) {
-            val location = response.header(LOCATION_KEY)
-            location?.let { ServerURLWrapper.setServerUrl(it) }
-            response.close()
-            val redirectReq = DynamicServerURLInterceptor.transformRequest(request)
-            response = chain.proceed(redirectReq)
-            redirects++
-        }
-        return response
-    }
+private const val SECTOMILISEC = 1000
 
-    companion object {
-        const val MaxRedirects = 20
+internal fun HttpClientConfig<*>.addKtorPlugins(
+    d2Configuration: D2Configuration,
+    authenticator: ParentAuthenticatorPlugin,
+) {
+    install(UserAgentPlugin) { this.d2Configuration = d2Configuration }
+    install(DynamicServerURLPlugin.instance)
+    install(ServerURLVersionRedirectionPlugin.instance)
+    install(PreventURLDecodePlugin.instance)
+    install(authenticator.instance)
+    install(HttpTimeout) {
```

Comment:
I am not sure if read/write timeouts are exactly equivalent. They might not have a direct equivalent in Ktor.

Maybe we can leave the timeout configuration in OkHttp because the D2Configuration object will be different for Android and iOS. In Ktor, we could apply the maximum value of them, just to be sure we don't set a lower timeout.

Reply:
That was the best equivalent I could find, but I agree that we can keep it as it was.